ignite/contrib/handlers/time_profilers.py

```diff
@@ -437,3 +438,299 @@ def odict_to_str(d: Mapping):
+            # TODO: add threshold filtering for handlers with event filter
+            # if (
+            #     isinstance(event_name, CallableEventWithFilter)
+            #     and event_name.filter != CallableEventWithFilter.default_event_filter
+            # ):
+            self._event_handlers_timer.reset()
+            func, args, kwargs = handler
+            func(*args, **kwargs)
+            t = self._event_handlers_timer.value()
+            hname = self._get_callable_name(func)
+            self.event_handlers_times[event][hname].append(t)
+
+        # required to revert back to original handler after profiling
+        _timeit_handler._profiler_original = handler
+        return _timeit_handler
+
+    # def _timeit_handler(self, handler: Callable, event: Events, *args, **kwargs):
+    #     # wrapper for profiling handlers attached to iteration events
+    #     self._event_handlers_timer.reset()
+    #     func, args, kwargs = handler
+    #     func(*args, **kwargs)
+    #     t = self._event_handlers_timer.value()
+    #     hname = self._get_callable_name(func)
+    #     self.event_handlers_times[event][hname].append(t)
+
+    def _timeit_processing(self):
+        # handler used for profiling processing times
+        t = self._processing_timer.value()
+        self.processing_times.append(t)
+
+    def _timeit_dataflow(self):
+        # handler used for profiling dataflow times
+        t = self._dataflow_timer.value()
+        self.dataflow_times.append(t)
+
+    def _reset(self, event_handlers_names):
+        # reset the variables used for profiling
+        self.dataflow_times = []
+        self.processing_times = []
+
+        self.event_handlers_times = {e: {h: [] for h in event_handlers_names[e]} for e in event_handlers_names}
+
+    @staticmethod
+    def _is_internal_handler(handler: Callable):
+        # checks whether the handler is internal
+        return "HandlersTimeProfiler." in repr(handler)
+
+    def _detach_profiler_handlers(self, engine):
+        # reverts handlers to original handlers
+        for e in engine._event_handlers:
+            for i, h in enumerate(engine._event_handlers[e]):
+                if hasattr(h[0], "_profiler_original"):
+                    engine._event_handlers[e][i] = h[0]._profiler_original
+
+    def _as_first_started(self, engine):
+
+        self.event_handlers_names = {
+            e: [
+                h.__qualname__ if hasattr(h, "__qualname__") else h.__class__.__name__
+                for (h, _, _) in engine._event_handlers[e]
+                if not self._is_internal_handler(h)
+            ]
+            for e in engine._allowed_events
+        }
+
+        self._reset(self.event_handlers_names)
+
+        for e in engine._allowed_events:
+            for i, h in enumerate(engine._event_handlers[e]):
+                if not self._is_internal_handler(h):
+                    engine._event_handlers[e][i] = (self._timeit_handler_wrapper(h), (e,), {})
```

Comment:
I wonder if we could only wrap handler function instead of replacing whole bundle as you do. And it would be good to use `functools.wraps` to decorate `_timeit_handler_wrapper` such that if user inspects handlers, he/she still sees its own handlers.

Reply:
Wrapper just needs event name but I think we can use `Engine.last_event_name` for that purpose as we get engine as first argument anyway. So the wrapper will look as follows with both the changes:
```python
def _timeit_handler_wrapper(self, handler: Callable):
    @functools.wraps(handler)
    def _timeit_handler(engine: Engine, *args: Any, **kwargs: Any):
        self._event_handlers_timer.reset()
        handler(engine, *args, **kwargs)
        t = self._event_handlers_timer.value()
        hname = self._get_callable_name(handler)
        self.event_handlers_times[engine.last_event_name][hname].append(t)

    # required to revert back to original handler after profiling
    setattr(_timeit_handler, "_profiler_original", handler)
    return _timeit_handler
```

Reply:
Oh sorry, this won't work. We have handler bundle as type tuple here so we won't be able to replace the item in that. I guess we will have to replace whole handler tuple.

Reply:
Yes, that's right, we have handler bundle as `(handler, args, kwargs)`. I was thinking to make the following replacement:
```
(handler, args, kwargs) -> (wrapped_handler, args, kwargs)
```
with `wrapped_handler = create_wrapped_handler(handler, event_name)`.

I'm not sure if it would work as you did but would be more simple to read the code ... Let's see.

Reply:
I just pushed the code to only change/wrap function inside bundle/tuple instead of replacing whole bundle. It still requires recreating bundle with same args and kwargs because of how Tuple works. It does seem more clean now though. Let me know if this looks good.

Reply:
@harsh8398 thanks ! Looks good now. We can even reduce it a bit like that:
```python
        for e in engine._allowed_events:
            for i, (func, args, kwargs) in enumerate(engine._event_handlers[e]):
                if not self._is_internal_handler(func):
                    engine._event_handlers[e][i] = (self._create_wrapped_handler(func, e), args, kwargs)
```